cabot/cabotapp/alert.py

```diff
@@ -23,17 +24,57 @@ def send_alert(self, service, users, duty_officers):
         """
         return True
 
+    def send_test_alert(self, user):
+        """
+        Send a test alert when the user requests it (to make sure config is valid).
+
+        The default implementation creates a fake Service and StatusCheck and calls the normal send_alert().
+        Note that the service/check may not be fully configured (e.g. invalid primary key, no HipChat room ID, ...).
+        :param user: django user
+        :return: nothing, raise exceptions on error
+        """
+        # local import because models.py depends on alert.py :\
+        from cabot.cabotapp.models import CheckGroupMixin
+
+        check_mock = Mock()
+        check_mock.configure_mock(id=0, pk=0, name='Alert Testing Check', active=True)
+
+        service_mock = Mock()
+        service_mock.configure_mock(id=0, pk=0, name='Alert Testing Service', alerts_enabled=True,
+                                    CRITICAL_STATUS=CheckGroupMixin.CRITICAL_STATUS,  # hipchat plugin uses this
+                                    status_checks=[check_mock], recent_snapshots=[],
+                                    overall_status=CheckGroupMixin.CRITICAL_STATUS,
+                                    active_status_checks=lambda: [check_mock],
+                                    all_passing_checks=lambda: [], all_failing_checks=lambda: [check_mock],
+                                    hipchat_instance=None, hipchat_room_id=None)
```

Comment:
Hmm if we don't set a hipchat instance or hipchat room id Hipchat alerts won't work...I'm fine with not having a Hipchat (or Pagerduty) test though, I think people are primarily concerned about phone call alerts working properly (would be good to not show the option in the UI then though)

Reply:
I think HipChat is supposed to fall back to some env vars for url/api key/room if anything is not set, so I was relying on that behavior for test alerts.

Maybe HipChat/PagerDuty etc should reimplement `send_test_alert` to deal with this - in HipChat's case, maybe send a DM to the user instead of posting to a room? The downside is this wouldn't test the same code path as a normal alert.

Also, there's currently no way for a plugin to signal that it doesn't support test alerts. Maybe I should make this default implementation opt-in?

Reply:
For now, I've moved the mock setup into a helper function. If we want to do test alerts with a mock for HipChat, etc., we can call the helper and add any plugin-specific fields in the plugin. Example:

```python
class HipChatAlertPlugin(AlertPlugin):
  def send_test_alert(self, user):
    service = create_failing_service_mock()
    service.configure_mock(hipchat_instance=..., hipchat_room_id=...)
    self.send_alert(service, [], [user])
```

(works because `Mock.configure_mock()` is additive)